File: csharp/P0001_TwoSum/Solutions/GroupedIndexMap.cs
// Time complexity: O(n)
// Space complexity: O(n)

namespace P0001_TwoSum.Solutions.GroupedIndexMap;

public class Solution
{
    public int[] TwoSum(int[] nums, int target)
    {
        // Build grouped index map
        var seen = new Dictionary<int, List<int>>();

        for (int i = 0; i < nums.Length; i++)
        {
            if (!seen.ContainsKey(nums[i]))
                seen[nums[i]] = new List<int>();
            
            seen[nums[i]].Add(i);
        }

        // Locate complement in index map
        for (int i = 0; i < nums.Length; i++)
        {
            int complement = target - nums[i];

            if (seen.ContainsKey(complement))
            {
                foreach (int j in seen[complement])
                    if (j != i)
                        return [i, j];
            }
        }
        
        // This should be unreachable
        return [-1];
    }
}

File: rust/p0001_two_sum/src/grouped_index_map.rs
// Time complexity: O(n)
// Space complexity: O(n)

use std::collections::HashMap;

pub struct Solution;

impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        // Build grouped index map.
        let mut seen: HashMap<i32, Vec<usize>> = HashMap::new();

        for (i, num) in nums.iter().enumerate() {
            seen.entry(*num).or_insert_with(Vec::new).push(i);
        }

        // Locate complement in grouped index map
        for (i, num) in nums.iter().enumerate() {
            let complement = target - num;

            if let Some(indices) = seen.get(&complement) {
                for &j in indices {
                    if j != i {
                        return vec![i as i32, j as i32];
                    }
                }
            }
        }

        unreachable!("This should be unreachable");
    }
}
